src/platform-fru-detect.cpp

```diff
@@ -20,11 +20,22 @@ int main(void)
 {
     PlatformManager pm;
 
-    Rainier0z rainier0z;
-    rainier0z.enrollWith(pm);
+    sdbusplus::bus::bus dbus = sdbusplus::bus::new_default();
+    InventoryManager inventory(dbus);
+    PublishWhenPresentInventoryDecorator decoratedInventory(&inventory);
+    std::unique_ptr<Rainier0z> rainier0z;
+    std::unique_ptr<Rainier1z> rainier1z;
 
-    Rainier1z rainier1z;
-    rainier1z.enrollWith(pm);
+    if (Rainier0z::isPresent(pm.getPlatformModel()))
```

Comment:
I'm not entirely sold on this kind of testing as it's the kind of behaviour that's meant to be implemented in the PlatforManager Further, conditionally instantiating different platform classes was what I was trying to avoid, I was hoping to instantiate them all and choose at runtime which one we need. Avoiding this helps keep the code a bit more linear, and the object lifetimes easier to understand.

More broadly, with dependency injection, I have been trying to keep ownership tied to stack frames. I feel delegating ownership to `std::unique_ptr`s that may or may not be populated makes it harder to follow what's going on.

Reply:
I guess I don't understand the desire to create a Platform object for the wrong system.  Really, PlatformManager should probably just hide the platforms within it so main doesn't even need to care and just create that correct one as necessary.

I could tell you wanted everything on the stack, but that doesn't work as well for match callbacks.

Technically I could have had each Flett register for its own propertiesChanged signals, but that gets tougher for cable cards.  Those can be added at standby and also probably at runtime, so then that would open up a whole new path that would have to be tested - adding new matches for Present and maybe Model, having to hang them somewhere, and  creating those objects appropriately.  The way I did it made hot adds for those an uninteresting case.

But anyway, I'll welcome your prototyping.  I really want to get this done this year so I will use whichever object model you want, just keep in mind my cable card hot add case above which I'd rather not have to add explicit support for.

Reply:
> I guess I don't understand the desire to create a Platform object for the wrong system. Really, PlatformManager should probably just hide the platforms within it so main doesn't even need to care and just create that correct one as necessary.

So there are a few inter-related issues here. The first is the principle of "keeping things that are together together" - an implementation of Platform should hold the information necessary to determine whether the platform we're running on matches the one it implements support for. In this case it's the set of `model` strings that come from the associated devicetrees.

From there we have a chicken-and-egg problem of we don't know what platform we're running on until a Platform instance tells us whether we're running on something that it's compatible with. This is done by extracting the `model` string from the devicetree and mapping that to a Platform instance, so we need the PlatformManager populated with all the supported platform instances for that mapping to correctly succeed or fail.

The last philosophical piece is dependency injection, or more to the point not having PlatformManager instantiate the Platform instances itself. This aids testing of PlatformManager functionality (not that there are any tests yet, but not taking this approach would preclude adding any in a clean fashion).

> Technically I could have had each Flett register for its own propertiesChanged signals, but that gets tougher for cable cards. Those can be added at standby and also probably at runtime, so then that would open up a whole new path that would have to be tested - adding new matches for Present and maybe Model, having to hang them somewhere, and creating those objects appropriately. The way I did it made hot adds for those an uninteresting case.

Thanks for the explanation! Now I can consider the trade-offs.

> But anyway, I'll welcome your prototyping. I really want to get this done this year so I will use whichever object model you want, just keep in mind my cable card hot add case above which I'd rather not have to add explicit support for.

I need this support to fix the drives reporting -17C temperatures, so it's currently on the top of the pile of things I'm looking at. I was hoping to push some patches before you returned from the break but got tied down with other work and personal things :(

Reply:
@spinler here's a WIP of what I was thinking for property change events: https://github.com/amboar/platform-fru-detect/commit/730a81da306079aaaea05e7162a71a1ef4c0920d

Note that I haven't implemented anything that registers handlers just yet, it's just a change that provides the infrastructure. The idea from here is that the `std::weak_ptr<PropertiesChangedListener>` that you get back from `Inventory::addPropertiesChangedListener()` is just a cookie that you can use to delete the listener later either in the class destructor or `unplug()` if it's a Device (`std::weak_ptr()` to indicate you don't own it, `PropertiesChangedListener` class with no public members or methods to prevent you from doing anything with it other than destroy it). So you'd do something like:

```
class MyDevice : public Device
{
  ...
  private:
    std::set<std::weak_ptr<PropertiesChangedListener>> listeners;
    Inventory* inventory;
}

void MyDevice::plug()
{
  auto mil = inventory->addPropertiesChangedListener("/my/item/path", "my.interface",
    [this](PropertiesChanged&& pc) mutable {
      pc.read(...);
      if (...)
      {
        this->foo();
      }
    });
  listeners.insert(mil);
  ...
}

void MyDevice::unplug()
{
  for (auto& listener  : listeners)
  {
    inventory->removePropertiesChangedListener(listener);
  }
}
```

I said above I needed this for fixing timing of polling the drive temperatures - I've changed my mind on that. I'm going to go poke at entity-manager a bit to solve some related problems there.

I was thinking though, perhaps what we need are `Exposes` clauses in the EM configs that are gated by things like power state? What do you think?

Reply:
FWIW, the Factory design pattern is made to create objects based on some criteria.  It's been around for decades.

Reply:
I have to come up with an 'upstreamable'  entity-manager design for this because it always relies on interfaces managed outside of entity-manager, which is a no-no.

Reply:
> FWIW, the Factory design pattern is made to create objects based on some criteria. It's been around for decades.

I am aware of it. It's a trade-off between that and the std::unique_ptr, and I'd rather avoid the std::unique_ptr stuff as I explained above. I think it's possible to achieve that, hence more prototyping for what you need here.